mobile_config/src/client/mod.rs

```diff
@@ -0,0 +1,100 @@
+use crate::gateway_info;
+use file_store::traits::MsgVerify;
+use futures::stream::{self, StreamExt};
+use helium_crypto::{Keypair, PublicKey, PublicKeyBinary, Sign};
+use helium_proto::{
+    services::{mobile_config, Channel},
+    Message,
+};
+use std::sync::Arc;
+
+mod settings;
+pub use settings::Settings;
+
+#[derive(thiserror::Error, Debug)]
+pub enum ClientError {
+    #[error("error signing request {0}")]
+    SigningError(#[from] helium_crypto::Error),
+    #[error("grpc error response {0}")]
+    GrpcError(#[from] tonic::Status),
+    #[error("error verifying response signature {0}")]
+    VerificationError(#[from] file_store::Error),
+}
+
+#[derive(Clone, Debug)]
+pub struct Client {
+    pub client: mobile_config::GatewayClient<Channel>,
+    signing_key: Arc<Keypair>,
+    config_pubkey: PublicKey,
+    batch_size: u32,
+}
+
+impl Client {
+    pub fn from_settings(settings: &Settings) -> Result<Self, Box<helium_crypto::Error>> {
+        Ok(Self {
+            client: settings.connect(),
+            signing_key: settings.signing_keypair()?,
+            config_pubkey: settings.config_pubkey()?,
+            batch_size: settings.batch_size,
+        })
+    }
+}
+
+#[async_trait::async_trait]
+impl gateway_info::GatewayInfoResolver for Client {
+    type Error = ClientError;
+
+    async fn resolve_gateway_info(
+        &mut self,
+        address: &PublicKeyBinary,
+    ) -> Result<Option<gateway_info::GatewayInfo>, Self::Error> {
```

Comment:
Does the return type here need to be a `Result<Option<GatewayInfo>>`? Can it just be `Result<GatewayInfo>` instead and handled accordingly downstream?

Reply:
I believe this is a valid return type for general use and it’s up to the caller to determine how to interpret that. It mirrors the retrieval of the record from the helius database in that regard in that we allow a successful request return a None when the gateway is not found on chain but we can error for other failure scenarios that could be recovered from like a malformed pubkey request argument or an unauthorized request.